velox/expression/tests/ExpressionVerifier.cpp

```diff
@@ -0,0 +1,290 @@
+        execCtx_, &exprSetSimplified, rowVector.get());
+
+    try {
+      exprSetSimplified.eval(rows, evalCtxSimplified, simplifiedEvalResult);
+    } catch (...) {
+      if (!canThrow) {
+        LOG(ERROR)
+            << "Simplified eval wasn't supposed to throw, but it did. Aborting.";
+        throw;
+      }
+      exceptionSimplifiedPtr = std::current_exception();
+    }
+  } catch (...) {
+    exceptionSimplifiedPtr = std::current_exception();
+  }
+
+  try {
+    // Compare results or exceptions (if any). Fail is anything is different.
+    if (exceptionCommonPtr || exceptionSimplifiedPtr) {
+      // Throws in case exceptions are not compatible. If they are compatible,
+      // return false to signal that the expression failed.
+      compareExceptions(exceptionCommonPtr, exceptionSimplifiedPtr);
+      return false;
+    } else {
+      // Throws in case output is different.
+      compareVectors(commonEvalResult.front(), simplifiedEvalResult.front());
+    }
+  } catch (...) {
+    if (!options_.reproPersistPath.empty()) {
+      persistReproInfo(rowVector, copiedResult, sql);
+    }
+    throw;
+  }
+  return true;
+}
+
+void ExpressionVerifier::persistReproInfo(
+    const VectorPtr& inputVector,
+    const VectorPtr& resultVector,
+    const std::string& sql) {
+  std::string inputPath;
+  std::string resultPath;
+  std::string sqlPath;
+
+  const auto basePath = options_.reproPersistPath.c_str();
+  // Saving input vector
+  auto inputPathOpt = generateFilePath(basePath, kVectorPathPrefix.data());
+  if (!inputPathOpt.has_value()) {
+    inputPath = "Failed to create file for saving input vector.";
+  } else {
+    inputPath = inputPathOpt.value();
+    try {
+      saveVectorToFile(inputVector.get(), inputPath.c_str());
+    } catch (std::exception& e) {
+      inputPath = e.what();
+    }
+  }
+
+  // Saving result vector
+  if (resultVector) {
+    auto resultPathOpt = generateFilePath(basePath, kVectorPathPrefix.data());
+    if (!resultPathOpt.has_value()) {
+      resultPath = "Failed to create file for saving result vector.";
+    } else {
+      resultPath = resultPathOpt.value();
+      try {
+        saveVectorToFile(resultVector.get(), resultPath.c_str());
+      } catch (std::exception& e) {
+        resultPath = e.what();
+      }
+    }
+  }
+
+  // Saving sql
+  auto sqlPathOpt = generateFilePath(basePath, kStringPathPrefix.data());
+  if (!sqlPathOpt.has_value()) {
+    sqlPath = "Failed to create file for saving SQL.";
+  } else {
+    sqlPath = sqlPathOpt.value();
+    try {
+      saveStringToFile(sql, sqlPath.c_str());
+    } catch (std::exception& e) {
+      resultPath = e.what();
+    }
+  }
+
+  std::stringstream ss;
+  ss << "Persisted input at '" << inputPath;
+  if (resultVector) {
+    ss << "' and result at '" << resultPath;
+  }
+  ss << "' and sql at '" << sqlPath << "'";
+  LOG(INFO) << ss.str();
+}
+
+
+void ExpressionVerifier::printRowVector(const RowVectorPtr& rowVector) {
```

Comment:
this method could be a free function in the anonymous namespace in the .cpp file